src/network/ares_dns_resolver/ares_library_handle.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+namespace lightstep {
+/**
+ * Manages initialization of the c-ares library.
+ */
+class AresLibraryHandle {
+ public:
+  AresLibraryHandle(const AresLibraryHandle&) = delete;
+  AresLibraryHandle(AresLibraryHandle&&) = delete;
+
+  AresLibraryHandle& operator=(const AresLibraryHandle&) = delete;
+  AresLibraryHandle& operator=(AresLibraryHandle&&) = delete;
+
+  /**
+   * @return if the libevent library initialization was successful.
+   */
+  bool initialized() const noexcept;
+
+  // We use a global variable to initialize the ares library because
+  // ares_library_init is not thread-safe and must be called before other
+  // threads are started.
+  //
+  // See https://c-ares.haxx.se/ares_library_init.html
+  static const AresLibraryHandle Instance;
```

Comment:
Is there a danger of the destructor being called at exit? I remember this sort of trouble being a real thing. Maybe this is safe, if so please comment.

Reply:
Good point. There's a lifetime issue, especially if the tracer's lifetime is also bound to a global.

I modified so that the global variable is now an `std::shared_ptr<const AresLibraryHandle>` and the AresDnsResolver stores a copy of the shared pointer to ensure the AresLibraryHandle isn't destroyed while the dns resolver is still active.